pmd-java/src/main/java/net/sourceforge/pmd/lang/java/ast/internal/ImportWrapper.java

```diff
@@ -8,57 +8,78 @@
+import net.sourceforge.pmd.RuleContext;
 import net.sourceforge.pmd.lang.ast.Node;
 import net.sourceforge.pmd.lang.java.ast.ASTImportDeclaration;
 
 /**
  * Helper class to analyze {@link ASTImportDeclaration}s.
  */
 public class ImportWrapper {
-    private ASTImportDeclaration node;
-    private String name;
-    private String fullname;
-    private boolean isStaticDemand;
-    private Set<String> allDemands = new HashSet<>();
+    private static final Logger LOG = Logger.getLogger(ImportWrapper.class.getName());
+
+    private final ASTImportDeclaration node;
+    private final String name;
+    private final String fullname;
+    private final boolean isStaticDemand;
+    private final Set<String> allStaticDemands;
 
     public ImportWrapper(String fullname, String name) {
         this(fullname, name, null);
     }
 
     public ImportWrapper(String fullname, String name, ASTImportDeclaration node) {
-        this(fullname, name, node, false);
+        this(fullname, name, node, false, null);
     }
 
-    public ImportWrapper(String fullname, String name, ASTImportDeclaration node, boolean isStaticDemand) {
+    public ImportWrapper(String fullname, String name, ASTImportDeclaration node, boolean isStaticDemand, RuleContext ctx) {
         this.fullname = fullname;
         this.name = name;
         this.node = node;
         this.isStaticDemand = isStaticDemand;
+        this.allStaticDemands = collectStaticFieldsAndMethods(node, ctx);
+
+    }
+
+    /**
+     * @param node
+     */
+    private Set<String> collectStaticFieldsAndMethods(ASTImportDeclaration node, RuleContext ctx) {
+        if (!this.isStaticDemand || node == null || node.getType() == null) {
+            return Collections.emptySet();
+        }
 
-        if (node != null && node.getType() != null) {
+        try {
+            Set<String> names = new HashSet<>();
             Class<?> type = node.getType();
-            for (Method m : type.getMethods()) {
-                allDemands.add(m.getName());
-            }
-            for (Field f : type.getFields()) {
-                allDemands.add(f.getName());
-            }
-            // also consider static fields, that are not public
+            // consider static fields, public and non-public
             for (Field f : type.getDeclaredFields()) {
                 if (Modifier.isStatic(f.getModifiers())) {
-                    allDemands.add(f.getName());
+                    names.add(f.getName());
                 }
             }
             // and methods, too
             for (Method m : type.getDeclaredMethods()) {
                 if (Modifier.isStatic(m.getModifiers())) {
-                    allDemands.add(m.getName());
+                    names.add(m.getName());
                 }
             }
+            return names;
+        } catch (LinkageError e) {
+            if (ctx != null) {
+                ctx.getReport().addError(new ProcessingError(e, String.valueOf(ctx.getSourceCodeFile())));
```

Comment:
I'm not sure adding a processing error is appropriate here. I don't think we currently log missing classes to the report right? And for now, processing errors indicate errors in PMD code, the kind of errors people should report as bugs. Currently, a processing error may very well mean that the file was not processed by rules at all, and so, violations may be missing from the report. Whereas missing dependencies is more akin to a warning, ie, we can continue analysis (and we do), but results of some rules may be somewhat inaccurate. And users shouldn't think this is a bug on our side.

I would say a log call would be more appropriate, as it's possible to disable it, and it makes it clearer that this is just a warning. In PMD 7 this will be handled by the semantic checker, which is also akin to a logger, and won't touch the report.

Reply:
Yes, you're right. This logging is something different than a processing error (and in other places, we silently ignore linkagerrors).

However, I still think, that such issues/warnings/problems should be part of the report, as it might indicate a incorrect/incomplete configuration when executing PMD. But that's for PMD 7 then.

I'll change that to a LOG.debug call. I'm not sure, how much duplicates we would log otherwise, and that could be annoying as well. See also #1371.